src/MultipartBody.cs

```diff
@@ -0,0 +1,158 @@
+    /// <param name="partName">The name of the part.</param>
+    /// <returns>The value of the part.</returns>
+    public T? GetPartValue<T>(string partName)
+    {
+        if(string.IsNullOrEmpty(partName))
+        {
+            throw new ArgumentNullException(nameof(partName));
+        }
+        if(_parts.TryGetValue(partName, out var value))
+        {
+            return (T)value.Item2;
+        }
+        return default;
+    }
+    /// <summary>
+    /// Removes a part from the multipart body.
+    /// </summary>
+    /// <param name="partName">The name of the part.</param>
+    /// <returns>True if the part was removed, false otherwise.</returns>   
+    public bool RemovePart(string partName)
+    {
+        if(string.IsNullOrEmpty(partName))
+        {
+            throw new ArgumentNullException(nameof(partName));
+        }
+        return _parts.Remove(partName);
+    }
+    private readonly Dictionary<string, Tuple<string, object>> _parts = new Dictionary<string, Tuple<string, object>>(StringComparer.OrdinalIgnoreCase);
+    /// <inheritdoc />
+    public IDictionary<string, Action<IParseNode>> GetFieldDeserializers() => throw new NotImplementedException();
+    /// <inheritdoc />
+    public void Serialize(ISerializationWriter writer)
+    {
+        if(writer == null)
+        {
+            throw new ArgumentNullException(nameof(writer));
+        }
+        if(RequestAdapter?.SerializationWriterFactory == null)
+        {
+            throw new InvalidOperationException(nameof(RequestAdapter.SerializationWriterFactory));
+        }
+        if(!_parts.Any())
+        {
+            throw new InvalidOperationException("No parts to serialize");
+        }
+        var first = true;
+        foreach(var part in _parts)
+        {
+            try
+            {
+                if(first)
+                    first = false;
+                else
+                    writer.WriteStringValue(string.Empty, string.Empty);
+
+                writer.WriteStringValue(string.Empty, $"--{Boundary}");
+                writer.WriteStringValue("Content-Type", $"{part.Value.Item1}");
+                writer.WriteStringValue("Content-Disposition", $"form-data; name=\"{part.Key}\"");
+                writer.WriteStringValue(string.Empty, string.Empty);
+                if(part.Value.Item2 is IParsable parsable)
+                {
+                    using var partWriter = RequestAdapter.SerializationWriterFactory.GetSerializationWriter(part.Value.Item1);
+                    partWriter.WriteObjectValue(string.Empty, parsable);
+                    using var partContent = partWriter.GetSerializedContent();
+                    if(partContent.CanSeek)
+                        partContent.Seek(0, SeekOrigin.Begin);
+                    using var ms = new MemoryStream();
+                    partContent.CopyTo(ms);
+                    writer.WriteByteArrayValue(string.Empty, ms.ToArray());
```

Comment:
Since there is a good chance that the returned stream is already a memory stream, 
- https://github.com/microsoft/kiota-serialization-json-dotnet/blob/7cbe39c2f548b59789dfe6b95fc5e799c0e7417b/src/JsonSerializationWriter.cs#L24
- https://github.com/microsoft/kiota-serialization-text-dotnet/blob/2f18bd75c07b59f41af401937b715ccb83e4c02d/src/TextSerializationWriter.cs#L16
- https://github.com/microsoft/kiota-serialization-form-dotnet/blob/9009d0f9fbbe91876a09a6038243dc499c05103d/src/FormSerializationWriter.cs#L26

We could maybe avoid the copy/new stream by doing something like this?

```cs
if(partContent is MemoryStream orginalMemoryStream)
    writer.WriteByteArrayValue(string.Empty, orginalMemoryStream.ToArray());
else
{
    using var ms = new MemoryStream();
    partContent.CopyTo(ms);
    writer.WriteByteArrayValue(string.Empty, ms.ToArray());
}
```

Reply:
good idea, done